algebra2/gaussian_elimination/gaussian_elimination.hpp

```diff
@@ -20,267 +22,322 @@ namespace algorithms::gaussian_elimination {
+              std::size_t row_j) {
         T temp;
-        for(int i = 0; i < m.extent(1); i++){    
+        for (std::size_t i = 0; i < m.number_of_columns(); i++) {
             temp = m[row_i, i];
             m[row_i, i] = m[row_j, i];
-            m[row_j, i] = temp; 
+            m[row_j, i] = temp;
         }
     }
 
+
+    template <typename T>
+    struct is_fraction : std::false_type {};
+
+
+    template <std::integral I>
+    struct is_fraction<fraction<I>> : std::true_type {};
+
+
+    template <typename T>
+    constexpr bool is_fraction_v = is_fraction<std::remove_cvref_t<T>>::value;
+
+
+    template <typename T>
+    concept check_fraction = is_fraction_v<T>;
+
+
     /*
         description:
             converts vector of ints to vector of fractions
     */
     template <std::ranges::range R>
-    requires(std::integral<std::ranges::range_value_t<R>>)
-    auto convert_to_vector_of_fractions(R& v) {
-        return v | std::views::transform([](auto i) { return fraction{i, 1}; }) |
-            std::ranges::to<std::vector>();
+    requires(std::integral<std::ranges::range_value_t<
+                 R>>) auto convert_to_vector_of_fractions(R& v) {
+        if constexpr (check_fraction<std::ranges::range_value_t<R>>) {
+            return v;
+        } else {
+            return v | std::views::transform([](auto i) {
+                       return fraction{i, 1};
+                   }) |
+                   std::ranges::to<std::vector>();
+        }
     }
 
-    /*
-        description:
-            converts matrix of ints to vector of fractions
-    */
-    template <std::integral I, typename LP>
-    auto convert_to_matrix_of_fractions(::ranges::matrix_view<I, LP>& m)
-    {
-        auto fracs{convert_to_vector_of_fractions(m)};
-        ranges::matrix_view<fraction<I>, LP> matrix_of_fracs{fracs, m.extent(0), m.extent(1), LP{}};
-        return std::pair{std::move(fracs), matrix_of_fracs};
-    }
 
     /*
         description:
-            overload to avoid errors while converting matrix of fractions to a matrix of fractions
+            converts matrix of ints to vector of fractions
     */
     template <typename I, typename LP>
-    auto convert_to_matrix_of_fractions(::ranges::matrix_view<fraction<I>, LP>& m)
-    {
-        return std::pair{std::move(m | std::ranges::to<std::vector>()), m};
+    auto convert_to_matrix_of_fractions(::ranges::matrix_view<I, LP>& m) {
+        if constexpr (check_fraction<I>) {
+            std::vector<I> copy(m.begin(), m.end());
+            ::ranges::matrix_view<I, LP> mat{
+                copy, m.extent(0), m.extent(1), LP{}};
+            return std::pair{std::move(copy), mat};
+        } else {
+            auto fracs{convert_to_vector_of_fractions(m)};
+            ranges::matrix_view<fraction<I>, LP> matrix_of_fracs{
+                fracs, m.extent(0), m.extent(1), LP{}};
+            return std::pair{std::move(fracs), matrix_of_fracs};
+        }
```

Comment:
move to matrix.cxx and rename it (this function can be easily generalized - you just change type of matrix so it is matrix util sth like `auto to_type()`